src/webots/nodes/WbCadShape.cpp

```diff
@@ -165,10 +226,83 @@ void WbCadShape::updateUrl() {
       }
     }
 
+    const QString extension = completeUrl.mid(completeUrl.lastIndexOf('.') + 1).toLower();
+    if (extension == "obj" && WbUrl::isWeb(completeUrl)) {
+      // ensure any mtl referenced by the obj file are also downloaded
+      if (areMaterialAssetsAvailable(completeUrl)) {
+        mObjMaterials.clear();
+        // generate mapping between referenced files and cached files
+        QStringList rawMaterials = objMaterialList(completeUrl);
+        foreach (QString material, rawMaterials) {
+          QString adjustedUrl = generateMaterialUrl(material, completeUrl);
+          assert(WbNetwork::instance()->isCached(adjustedUrl));
+          if (!mObjMaterials.contains(material))
+            mObjMaterials.insert(material, adjustedUrl);
+        }
+      } else {
+        retrieveMaterials();
+        return;
+      }
+    }
+
     createWrenObjects();
   }
 }
 
+bool WbCadShape::areMaterialAssetsAvailable(const QString &url) {
+  QStringList rawMaterials = objMaterialList(url);  // note: 'dae' files will generate an empty list
+  foreach (QString material, rawMaterials) {
+    if (!WbNetwork::instance()->isCached(generateMaterialUrl(material, url)))
+      return false;
+  }
+  return true;
+}
+
+void WbCadShape::generateMaterialMap(const QString &url) {
+  QStringList rawMaterials = objMaterialList(url);
+  foreach (QString material, rawMaterials) {
+    QString adjustedUrl = generateMaterialUrl(material, url);
+    if (!WbNetwork::instance()->isCached(adjustedUrl)) {  // only generate the map when everything is available
+      mObjMaterials.clear();
+      warn(tr("Material asset '%1' is not available.").arg(adjustedUrl));
+      return;
+    }
+
+    if (!mObjMaterials.contains(material))
+      mObjMaterials.insert(material, WbNetwork::instance()->get(adjustedUrl));
+  }
+}
+
+QStringList WbCadShape::objMaterialList(const QString &url) {
+  assert(WbNetwork::instance()->isCached(url));  // should not search for materials if the obj file itself is not available
+  const QString extension = url.mid(url.lastIndexOf('.') + 1).toLower();
+  if (extension != "obj")
+    return QStringList();
+
+  QStringList materials;
+  QFile objFile(WbNetwork::instance()->get(url));
+  if (objFile.open(QIODevice::ReadOnly)) {
+    QString content = QString(objFile.readAll());
+    content = content.replace("\r\n", "\n");
+
+    QStringList lines = content.split('\n', Qt::SkipEmptyParts);
+    foreach (QString line, lines) {
+      const QString cleanLine = line.trimmed();
+      if (!cleanLine.startsWith("mtllib"))
+        continue;
+
+      materials = cleanLine.split(' ', Qt::SkipEmptyParts);
+      materials.removeFirst();  // first is "mtllib"
+      break;                    // only one occurrence of mtllib is allowed, so as soon as one is found the search can stop
```

Comment:
Is this a restriction that you want to introduce?
Because it is possible to have several `.mtl` in a `.obj` (https://en.wikipedia.org/wiki/Wavefront_.obj_file#Reference_materials)
I tested to have a `.obj` with two `mtllib` call with different `.mtl` file and I could load it in Blender and also in Webots (with local file)

Reply:
Then I must have misunderstood the [specs](http://paulbourke.net/dataformats/obj/):

> When you assign a material library using the Model program, only one map library per .obj file is allowed. You can assign multiple libraries using a text editor.

Something like `mtllib filename1.mtl filename2.mtl` is supported already. You tested something like:
```
mtllib filename1.mtl
mtllib filename2.mtl
```
and blender didn't complain? For instance if both files define the same materials but they are of different colors, which will it pick?

Either way I can add support for that second case, in the off-chance somebody has it

Reply:
and blender didn't complain?
Exactly and Webots (so assimp) handle it well as well (locally).
If you want to test it: https://github.com/BenjaminDeleze/sandbox/tree/master/meshes
I did not test to define two materials with the same name

Reply:
Then we should support it, thanks!